nameresolution/consul/consul.go

```diff
@@ -66,28 +65,142 @@ type healthInterface interface {
 }
 
 type resolver struct {
-	config resolverConfig
-	logger logger.Logger
-	client clientInterface
+	config   resolverConfig
+	logger   logger.Logger
+	client   clientInterface
+	registry registryInterface
+}
+
+type registryInterface interface {
+	get(service string) *registryEntry
+	expire(service string) // clears slice of instances
+	remove(service string) // removes entry from registry
+	addOrUpdate(service string, services []*consul.ServiceEntry)
+}
+
+type registry struct {
+	entries map[string]*registryEntry
+	mu      sync.RWMutex
+}
+
+type registryEntry struct {
+	services []*consul.ServiceEntry
+	mu       sync.RWMutex
+}
+
+func (r *registry) get(service string) *registryEntry {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+
+	return r.entries[service]
+}
+
+func (e *registryEntry) next() *consul.ServiceEntry {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	if len(e.services) == 0 {
+		return nil
+	}
+
+	return shuffle(e.services)[0]
+}
+
+func (r *resolver) getService(service string) (*consul.ServiceEntry, error) {
+	var services []*consul.ServiceEntry
+
+	if r.config.UseCache {
+		var entry *registryEntry
+
+		if entry = r.registry.get(service); entry != nil {
+			result := entry.next()
+
+			if result != nil {
+				return result, nil
+			}
+		} else {
+			r.watchService(service)
+		}
+	}
+
+	options := *r.config.QueryOptions
+	options.WaitHash = ""
+	options.WaitIndex = 0
+	services, _, err := r.client.Health().Service(service, "", true, &options)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to query healthy consul services: %w", err)
+	} else if len(services) == 0 {
+		return nil, fmt.Errorf("no healthy services found with AppID:%s", service)
+	}
+
+	return shuffle(services)[0], nil
+}
+
+func (r *registry) addOrUpdate(service string, services []*consul.ServiceEntry) {
+	var entry *registryEntry
+
+	// update
+	if entry = r.get(service); entry != nil {
+		entry.mu.Lock()
+		defer entry.mu.Unlock()
+
+		entry.services = services
+
+		return
+	}
+
+	// add
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	r.entries[service] = &registryEntry{
+		services: services,
+	}
+}
+
+func (r *registry) remove(service string) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if _, ok := r.entries[service]; !ok {
+		return
+	}
+
+	delete(r.entries, service)
+}
+
+func (r *registry) expire(service string) {
```

Comment:
What is the different between remove and expire?

Reply:
`remove()` will remove the service from the registry and is only called in the case of a watcher panic, any subsequent call will add the service again and start a watcher. `expire()` will clear all nodes/entries from the cache for the given service, this is invoked when calls to the agent are erroring but the watcher is still active and will continue attempting calls to the agent on a backoff retry